src/librustc_mir/dataflow/impls/reaching_defs.rs

```diff
@@ -0,0 +1,414 @@
+/// `StatementKind::Assign`). However, a variable can also be implicitly defined by calling code if
+/// it is a function parameter or defined to be uninitialized via a call to
+/// `MaybeUninit::uninitialized`.
+///
+/// An example, using parentheses to denote each definition.
+///
+/// ```rust,no_run
+/// fn collatz_step(mut n: i32) -> i32 { // (1) Implicit definiton of function param
+///     if n % 2 == 0 {
+///         n /= 2                       // (2)
+///     } else {
+///         n = 3*n + 1                  // (3)
+///     }
+///
+///     // Only definitions (2) and (3) could possibly constitute the value of `n` at function exit.
+///     // (1), the `n` which was passed into the function, is killed by either (2) or (3)
+///     // along all paths to this point.
+///     n
+/// }
+/// ```
+///
+/// Not all programs are so simple; some care is required to handle projections. For field and
+/// array index projections, we cannot kill all previous definitions of the same variable, since
+/// only part of the variable was defined. If we encounter a deref projection on the left-hand side
+/// of an assignment, we don't even know *which* variable is being defined, since the pointer being
+/// dereferenced may be pointing anywhere. Such definitions are never killed. Finally, we must
+/// treat function calls as opaque: We can't know (without MIR inlining or explicit annotation)
+/// whether a callee mutates data behind a pointer. If the address of one of our locals is
+/// observable, it may be the target of such a mutation. A type-based analysis (e.g. does this
+/// function take any type containing a mutable reference as a parameter?) is insufficient, since
+/// raw pointers can be laundered through any integral type.
+///
+/// At the moment, the possible targets of a definition are tracked at the granularity of a
+/// `Local`, not a `MovePath` as is done in the initialized places analysis. That means that a
+/// definition like `x.y = 1` does not kill a previous assignment of `x.y`. This could be made more
+/// precise with more work; it is not a fundamental limitation of the analysis.
+///
+/// Most passes will want to know which definitions of a **specific** local reach a given point in
+/// the program. `ReachingDefinitions` does not provide this information; it can only provide all
+/// definitions of **any** local. However, a `ReachingDefinitions` analysis can be used to build a
+/// `UseDefChain`, which does provide this information.
+pub struct ReachingDefinitions {
+    all: IndexVec<DefIndex, Definition>,
+
+    /// All (direct) definitions of a local in the program.
+    for_local_direct: IndexVec<Local, Vec<DefIndex>>,
+
+    /// All definitions at the given location in the program.
+    at_location: FxHashMap<Location, SmallVec<[DefIndex; 4]>>,
+}
+
+fn is_call_with_side_effects(terminator: &mir::Terminator<'tcx>) -> bool {
+    if let mir::TerminatorKind::Call { .. } = terminator.kind {
```

Comment:
Maybe return false for pure intrinsics.

Reply:
Indeed. I was planning on doing this eventually, but I didn't want to spend time going through [`std::intrinsics`](https://doc.rust-lang.org/std/intrinsics/index.html). I'll at least add some machinery based on [`is_rustc_peek`](https://github.com/rust-lang/rust/blob/1301422a6c2e8916560b8cc2f0564f38d8858a75/src/librustc_mir/transform/rustc_peek.rs#L220) and a comment.

Reply:
`rustc_peek` has been special-cased. I also realized this was incorrect, since types with custom `Drop` impls can also have side effects. This has been fixed in the latest set of commits.